cdap-data-fabric/src/main/java/co/cask/cdap/data/dataset/SystemDatasetInstantiator.java

```
@@ -0,0 +1,90 @@
+ * can be properly cleaned up. For example, a {@link DirectoryClassLoaderProvider} will fetch dataset jars
+ * and unpack them in a local directory, which must be cleaned up after the dataset is no longer needed.
+ */
+public class SystemDatasetInstantiator implements Closeable {
+
+  private final DatasetFramework datasetFramework;
+  // provides classloaders to use for different dataset modules
+  private final DatasetClassLoaderProvider classLoaderProvider;
+  private final Iterable<? extends Id> owners;
+  private final ClassLoader parentClassLoader;
+
+  SystemDatasetInstantiator(DatasetFramework datasetFramework,
+                            @Nullable ClassLoader parentClassLoader,
+                            DatasetClassLoaderProvider classLoaderProvider,
+                            @Nullable Iterable<? extends Id> owners) {
+    this.owners = owners;
+    this.classLoaderProvider = classLoaderProvider;
+    this.datasetFramework = datasetFramework;
+    this.parentClassLoader = parentClassLoader == null ?
+      Objects.firstNonNull(Thread.currentThread().getContextClassLoader(), getClass().getClassLoader()) :
+      parentClassLoader;
+  }
+
+  public <T extends Dataset> T getDataset(Id.DatasetInstance datasetId)
+    throws DatasetInstantiationException {
+    return getDataset(datasetId, DatasetDefinition.NO_ARGUMENTS);
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T extends Dataset> T getDataset(Id.DatasetInstance datasetId, Map<String, String> arguments)
+    throws DatasetInstantiationException {
+
+    T dataset;
+    try {
+      if (!datasetFramework.hasInstance(datasetId)) {
+        throw new DatasetInstantiationException("Trying to access dataset that does not exist: " + datasetId);
+      }
+
+      dataset = datasetFramework.getDataset(datasetId, arguments, parentClassLoader, classLoaderProvider, owners);
+      if (dataset == null) {
```

can this return null? when would that happen? Shouldn't it rather throw an exception with the reason for the failure?